### ISSUE TITLE
To-D doesn't work in a Kamelet

### ISSUE THREAD

**Issue by oscerd** [apache/camel-kamelets#449]
With 1.5.0 the parameter substitution won't work in a to-d, like for example azure-storage-blob source and dropbox source.
We have some workaround, like setting an exchange property and then work through headers, but I think we need to address the situation.

**Comment by lburgazzoli** [comment]
```yaml
apiVersion: camel.apache.org/v1alpha1
kind: Kamelet
metadata:
  name: azure-storage-blob-source
spec:
  definition:
    title: "Azure Storage Blob Source"
    description: |-
      Consume Files from Azure Storage Blob.
    required:
      - period
      - accountName
      - containerName
      - accessKey
    type: object
    properties:
      period:
        title: Period between Polls
        description: The interval between fetches to the Azure Storage Container in milliseconds
        type: integer
        default: 10000
      accountName:
        title: Account Name
        description: The Azure Storage Blob account name.
        type: string
      containerName:
        title: Container Name
        description: The Azure Storage Blob container name.
        type: string
      accessKey:
        title: Access Key
        description: The Azure Storage Blob access Key.
        type: string
        format: password
        x-descriptors:
        - urn:alm:descriptor:com.tectonic.ui:password
  dependencies:
    - "camel:azure-storage-blob"
    - "camel:kamelet"
    - "camel:core"
    - "camel:jsonpath"
    - "camel:timer"
  flow:
    from:
      uri: "timer:azure-storage-blob-stream"
      parameters:
        period: "{{period}}"
      steps:
      - to:
          uri: "azure-storage-blob:{{accountName}}/{{containerName}}"
          parameters:
            operation: "listBlobs"
            accessKey: "{{accessKey}}"
      - split:
          jsonpath: "$.*"
          steps:
          - set-property:
              name: azureBlobName
              simple: ${body.name}
          - to-d: "azure-storage-blob:{{accountName}}/{{containerName}}?accessKey=RAW({{accessKey}})&operation=getBlob&blobName=${exchangeProperty.azureBlobName}"
          - to: "kamelet:sink"
```

@davsclaus I think the `to-d` here fails because when the real url is computed, the local properties are not more known.

**Comment by davsclaus** [comment]
@lburgazzoli yeah toD and recipientList etc that compute dynamic endpoint uris would be affected. Can you create a JIRA ticket

**Comment by davsclaus** [comment]
I created a JIRA
https://issues.apache.org/jira/browse/CAMEL-16823

**Comment by davsclaus** [comment]
If you can use headers then that is often better as this avoids creating dynamic endpoints which has higher overhead than static endpoints.

**Comment by oscerd** [comment]
For azure i solved in that way, there is still the Dropbox source kamelet and in that case the headers cannot be used because the component is not using the for producer operations

**Comment by davsclaus** [comment]
Okay a source with a "from" endpoint is not dynamic so there is no problem there.

**Comment by lburgazzoli** [comment]
wonder if the DefaultComponent should automatically resolve the placeholders when computing the remaining part of the uri that is used to invoke createEndpoint

**Comment by oscerd** [comment]
> Okay a source with a "from" endpoint is not dynamic so there is no problem there.

The source is done with a timer as from. Because there is no real consumer. So it's timer+producer. Like azure storage blob too

**Comment by davsclaus** [comment]
Okay so its specific the toD EIP that via the uri checks if the component is a special toD optimized (send dynamic aware) and this is causing this problem.

The route template do create the endpoint with the resolved parameters from the template, so that is correct. So fixing this in toD and this should work afterwards.

**Comment by davsclaus** [comment]
Fixed in Camel 3.12
